src/backends/torch/torchlib.cc

```diff
@@ -277,6 +277,66 @@ namespace dd
     return out_val;
   }
 
+  c10::IValue TorchModule::extract(std::vector<c10::IValue> source,
+                                   std::string extract_layer)
+  {
+    if (_graph) // native modules take only one tensor as input for now
+      return _graph->extract(to_tensor_safe(source[0]), extract_layer);
+    if (_native)
+      return _native->extract(to_tensor_safe(source[0]), extract_layer);
+    auto output = _traced->forward(source);
+    if (output.isTensorList())
+      {
+        auto elems = output.toTensorList();
+        source = std::vector<c10::IValue>(elems.begin(), elems.end());
+      }
+    else if (output.isTuple())
+      {
+        auto &elems = output.toTuple()->elements();
+        source = std::vector<c10::IValue>(elems.begin(), elems.end());
+      }
+    else
+      {
+        source = { output };
+      }
+    c10::IValue out_val = source.at(_classif_in);
+    if (_hidden_states)
+      {
+        // out_val is a tuple containing tensors of dimension n_batch *
+        // sequence_length * n_features We want a tensor of size n_batch *
+        // n_features from the last hidden state
+        auto &elems = out_val.toTuple()->elements();
+        out_val = elems.back().toTensor().slice(1, 0, 1).squeeze(1);
+      }
+    return out_val;
+  }
+
+  bool TorchModule::extractable(std::string extract_layer) const
+  {
+    if (_graph)
+      return _graph->extractable(extract_layer);
+    if (_native)
+      return _native->extractable(extract_layer);
+    if (_traced)
+      return extract_layer == "final";
```

Comment:
For my understanding, there's no other option in traced model ? Inner layers cannot be extracted ? cc @BynaryCobweb

Reply:
On traced models, we can only read the output tensors of the models, not the intermediate ones. So if we want to extract a layer, we must do it when we trace the model, and include it in the output of the model